[C10D] Push store scalability a bit further.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109218
* __->__ #109217

This is a bunch of small changes to improve store scalability:

- stagger client connection to avoid a stampede.
- warn if somaxconn is too small.
- increase the backlog to 16k.

Differential Revision: [D49238587](https://our.internmc.facebook.com/intern/diff/D49238587)

kumpera committed:
[C10D] Push store scalability a bit further.

This is a bunch of small changes to improve store scalability:

- stagger client connection to avoid a stampede.
- warn if somaxconn is too small.
- increase the backlog to 16k.

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49238587).

wconstab commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -299,11 +303,39 @@ TCPStore::TCPStore(std::string host, const TCPStoreOptions& opts)
+              *numWorkers_,
+              somaxconn);
+        }
+      } catch (std::exception& e) {
```
its tempting to catch all here to avoid adding a new source of error to the application.  but should we? can you just check the types of exceptions thrown by stoll and only catch those?

kumpera replied:
good point, it throws a bunch of exceptions, they all nail down to logic_error.

kumpera committed:
Update on "[C10D] Push store scalability a bit further."


This is a bunch of small changes to improve store scalability:

- stagger client connection to avoid a stampede.
- warn if somaxconn is too small.
- increase the backlog to 16k.

Differential Revision: [D49238587](https://our.internmc.facebook.com/intern/diff/D49238587)

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Push store scalability a bit further."


This is a bunch of small changes to improve store scalability:

- stagger client connection to avoid a stampede.
- warn if somaxconn is too small.
- increase the backlog to 16k.

Differential Revision: [D49238587](https://our.internmc.facebook.com/intern/diff/D49238587)

[ghstack-poisoned]

XilunWu reviewed (approved):
LGTM!

kumpera commented:
@pytorchmergebot  merge

kumpera commented:
@pytorchmergebot merge -f "error was on infra"
